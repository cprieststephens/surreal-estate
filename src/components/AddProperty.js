import React, { useState } from "react";
import "../styles/add-property.css";
import postProperty from "../requests/postProperty";
import Alert from "./Alert";

const AddProperty = () => {
  const initialState = {
    fields: {
      title: "",
      type: "Flat",
      bedrooms: 1,
      bathrooms: 1,
      price: 0,
      city: "Manchester",
      email: "",
    },
    alert: {
      message: "",
      isSuccess: false,
    },
  };

  const [fields, setFields] = useState(initialState.fields);
  const [alert, setAlert] = useState(initialState.alert);

  const handleFieldChange = (event) => {
    setFields({ ...fields, [event.target.name]: event.target.value });
  };

  const handleAddProperty = (event) => {
    event.preventDefault();
    setAlert({ message: "", isSuccess: false });
    postProperty(fields, setAlert);
  };

  return (
    <div className="add-property">
      <Alert message={alert.message} success={alert.isSuccess} />
      <form onSubmit={handleAddProperty} className="add-property__form">
        <div className="add-property__form-item">
          <label htmlFor="title">
            <span>Title: </span>
            <input
              className="add-property__free-text"
              type="text"
              placeholder="Flat available"
              id="title"
              name="title"
              value={fields.title}
              onChange={handleFieldChange}
            />
          </label>
        </div>
        <div className="add-property__form-item">
          <label htmlFor="type">
            <span>Type: </span>
            <select
              className="add-property__dropdown"
              id="type"
              name="type"
              value={fields.type}
              onChange={handleFieldChange}
            >
              <option value="Flat">Flat</option>
              <option value="Detached">Detached</option>
              <option value="Semi-Detached">Semi-Detached</option>
              <option value="Terraced">Terraced</option>
              <option value="End of Terrace">End of Terrace</option>
              <option value="Cottage">Cottage</option>
              <option value="Bungalow">Bungalow</option>
            </select>
          </label>
        </div>
        <div className="add-property__form-item">
          <label htmlFor="bedrooms">
            <span>Bedrooms: </span>
            <input
              className="add-property__number"
              type="number"
              id="bedrooms"
              name="bedrooms"
              value={fields.bedrooms}
              onChange={handleFieldChange}
            />
          </label>
        </div>
        <div className="add-property__form-item">
          <label htmlFor="bathrooms">
            <span>Bathrooms: </span>
            <input
              className="add-property__number"
              type="number"
              id="bathrooms"
              name="bathrooms"
              value={fields.bathrooms}
              onChange={handleFieldChange}
            />
          </label>
        </div>
        <div className="add-property__form-item">
          <label htmlFor="price">
            <span>Price: Â£ </span>
            <input
              className="add-property__price"
              type="number"
              id="price"
              name="price"
              value={fields.price}
              onChange={handleFieldChange}
            />
          </label>
        </div>
        <div className="add-property__form-item">
          <label htmlFor="city">
            <span>City: </span>
            <select
              className="add-property__dropdown"
              id="city"
              name="city"
              value={fields.city}
              onChange={handleFieldChange}
            >
              <option value="Manchester">Manchester</option>
              <option value="Leeds">Leeds</option>
              <option value="Sheffield">Sheffield</option>
              <option value="Liverpool">Liverpool</option>
            </select>
          </label>
        </div>
        <div className="add-property__form-item">
          <label htmlFor="email">
            <span>Email: </span>
            <input
              className="add-property__free-text"
              type="email"
              placeholder="jane.doe@email.co.uk"
              id="email"
              name="email"
              value={fields.email}
              onChange={handleFieldChange}
            />
          </label>
        </div>
        <button type="submit" className="add-property__button">
          Add
        </button>
      </form>
    </div>
  );
};

export default AddProperty;
