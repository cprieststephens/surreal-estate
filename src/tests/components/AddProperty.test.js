import React from "react";
import { render, screen } from "@testing-library/react";
import AddProperty from "../../components/AddProperty";

describe("AddProperty", () => {
  it("renders correctly", () => {
    const { asFragment } = render(<AddProperty />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the correct inputs", () => {
    render(<AddProperty />);
    const title = screen.getByLabelText(/title/i);
    const type = screen.getByLabelText(/type/i);
    const bedrooms = screen.getByLabelText(/bedrooms/i);
    const bathrooms = screen.getByLabelText(/bathrooms/i);
    const price = screen.getByLabelText(/price/i);
    const city = screen.getByLabelText(/city/i);
    const email = screen.getByLabelText(/email/i);

    expect(title).toBeInTheDocument();
    expect(type).toBeInTheDocument();
    expect(bedrooms).toBeInTheDocument();
    expect(bathrooms).toBeInTheDocument();
    expect(price).toBeInTheDocument();
    expect(city).toBeInTheDocument();
    expect(email).toBeInTheDocument();
  });

  it("renders a submit button", () => {
    render(<AddProperty />);
    const submitButton = screen.getByRole("button");

    expect(submitButton).toBeInTheDocument();
  });
});
