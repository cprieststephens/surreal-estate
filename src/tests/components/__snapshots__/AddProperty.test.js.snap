// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProperty renders correctly 1`] = `
<DocumentFragment>
  <div
    class="add-property"
  >
    <form
      class="add-property__form"
    >
      <div
        class="add-property__form-item"
      >
        <label
          for="title"
        >
          <span>
            Title: 
          </span>
          <input
            class="add-property__free-text"
            id="title"
            name="title"
            placeholder="Flat available"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="add-property__form-item"
      >
        <label
          for="type"
        >
          <span>
            Type: 
          </span>
          <select
            class="add-property__dropdown"
            id="type"
            name="type"
          >
            <option
              value="Flat"
            >
              Flat
            </option>
            <option
              value="Detached"
            >
              Detached
            </option>
            <option
              value="Semi-Detached"
            >
              Semi-Detached
            </option>
            <option
              value="Terraced"
            >
              Terraced
            </option>
            <option
              value="End of Terrace"
            >
              End of Terrace
            </option>
            <option
              value="Cottage"
            >
              Cottage
            </option>
            <option
              value="Bungalow"
            >
              Bungalow
            </option>
          </select>
        </label>
      </div>
      <div
        class="add-property__form-item"
      >
        <label
          for="bedrooms"
        >
          <span>
            Bedrooms: 
          </span>
          <input
            class="add-property__number"
            id="bedrooms"
            name="bedrooms"
            type="number"
            value="1"
          />
        </label>
      </div>
      <div
        class="add-property__form-item"
      >
        <label
          for="bathrooms"
        >
          <span>
            Bathrooms: 
          </span>
          <input
            class="add-property__number"
            id="bathrooms"
            name="bathrooms"
            type="number"
            value="1"
          />
        </label>
      </div>
      <div
        class="add-property__form-item"
      >
        <label
          for="price"
        >
          <span>
            Price: Â£ 
          </span>
          <input
            class="add-property__price"
            id="price"
            name="price"
            type="number"
            value="0"
          />
        </label>
      </div>
      <div
        class="add-property__form-item"
      >
        <label
          for="city"
        >
          <span>
            City: 
          </span>
          <select
            class="add-property__dropdown"
            id="city"
            name="city"
          >
            <option
              value="Manchester"
            >
              Manchester
            </option>
            <option
              value="Leeds"
            >
              Leeds
            </option>
            <option
              value="Sheffield"
            >
              Sheffield
            </option>
            <option
              value="Liverpool"
            >
              Liverpool
            </option>
          </select>
        </label>
      </div>
      <div
        class="add-property__form-item"
      >
        <label
          for="email"
        >
          <span>
            Email: 
          </span>
          <input
            class="add-property__free-text"
            id="email"
            name="email"
            placeholder="jane.doe@email.co.uk"
            type="email"
            value=""
          />
        </label>
      </div>
      <button
        class="add-property__button"
        type="submit"
      >
        Add
      </button>
    </form>
  </div>
</DocumentFragment>
`;
